[Rebase & FF] Standalone MM IPL fixes

kuqin12 opened the pull request:
## Description

This change cherry-picks 2 commits from EDK2 that could cause some platforms to not work with Standalone MM as expected.

The last commit is an add-on to resolve the feedback on this PR and will be taken to edk2.

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

This was tested on QEMU Q35 platform and booted to Windows desktop.

## Integration Instructions

N/A

kuqin12 committed:
[CHERRY-PICK] StandaloneMmPkg: StandaloneMmIplPei: Prevent Freeing Zero Pages

There could be scenarios where the HOB producer does not create any HOBs.
In such cases, the buffer intended to be freed will have zero pages.

This update addresses the issue that could cause assertions during
runtime by ensuring that buffers with zero pages are not freed.

Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit c72d6384347bc57109d75fd56f2a5cdb8a40d132)

kuqin12 committed:
[CHERRY-PICK] StandaloneMmPkg: StandaloneMmIplPei: Use MM access to open the regions

Current MM IPL in PEI phase does not open the MMRAM regions through MM
access PPI. This is causing some platforms like OVMF reading all `0xFF`s
when trying to relocate the Standalone MM core.

This change opens all the MMRAM regions provided by MM access PPI and
closes + locks the regions after initial MM foundation setup, when MM
Access PPI is available.

Platforms that require MM access PPI can inject depex through libraries.

Signed-off-by: Kun Qin <kun.qin@microsoft.com>
(cherry picked from commit 73b0b5edaef19beb08b04869bc2e0f6328c9e604)

makubacki commented on StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c:
```diff
@@ -680,7 +721,7 @@ ExecuteMmCoreFromMmram (
   Status = PeCoffLoaderGetImageInfo (&ImageContext);
   if (EFI_ERROR (Status)) {
-    return Status;
+    goto Done;
```
The function returns `Status`. At this point, `Status` is some error. Are you expecting it to be overwritten by the status returned from `MmAccess->Close()` and `MmAccess->Lock()` in the `Done` block?

kuqin12 replied:
A separate status is added to host returned results.

makubacki commented on StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c:
```diff
@@ -753,6 +795,44 @@ ExecuteMmCoreFromMmram (
+      }
+
+      //
+      // Print debug message that the MMRAM window is now closed.
```
```suggestion
      // Print debug message that the MMRAM window is now locked.
```

kuqin12 replied:
Updated.

makubacki commented on StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c:
```diff
@@ -753,6 +795,44 @@ ExecuteMmCoreFromMmram (
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_ERROR, "MM IPL failed to close MMRAM windows index %d - %r\n", Index, Status));
+        ASSERT (FALSE);
+        return Status;
```
Should this really return instead of attempting to close and lock any potentially remaining windows?

kuqin12 replied:
This is updated.

makubacki commented:
I know this is a cherry-pick but I had a few questions before finishing the review.

os-d commented on StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c:
```diff
@@ -756,6 +795,44 @@ ExecuteMmCoreFromMmram (
+      DEBUG ((DEBUG_INFO, "MM IPL closed MMRAM window index %d\n", Index));
+
+      //
+      // Lock the MMRAM (Note: Locking MMRAM may not be supported on all platforms)
```
If some platforms don't support locking, will they return an error here? In which case, should we explicitly check for that error code? e.g. if it is EFI_UNSUPPORTED. If it isn't uniform, then we can't always assert and return the status, right?

kuqin12 replied:
The idea (at least from Ray) is that those platforms should not produce MM access before IPL runs. In that case, this whole block will be skipped.

kuqin12 replied:
But a check is added.

kuqin12 committed:
StandaloneMmPkg: StandaloneMmIplPei: Handle return status from MmAccess

This change adds some logic to handle the MmAccess PPI failure and makes
effort to close and lock all regions when possible.

Signed-off-by: Kun Qin <kuqin12@gmail.com>
